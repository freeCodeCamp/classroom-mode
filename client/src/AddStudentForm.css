.class-table-button-container {
  background-color: #E0E0E0;
  height: 60px;
  display: flex;
  align-items: center;
}

.open-modal {
  margin-left: 120px;
}

#student-count {
  margin-left: auto;
  margin-right: 120px;
}