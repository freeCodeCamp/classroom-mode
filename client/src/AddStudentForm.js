import React, { Component } from 'react';
import { Button,
         Modal,
         FormGroup,
         ControlLabel,
         FormControl } from 'react-bootstrap';


class AddStudentForm extends Component {
  constructor(props){
    super(props);
    this.state = {
      showModal: false,
      name: '',
      username: '',
      email: '',
      notes: '',
      errors: []
    };
    this.open = this.open.bind(this);
    this.close = this.close.bind(this);
    this.submit = this.submit.bind(this);
    this.handleChange = this.handleChange.bind(this);
  }

  close() {
    this.setState({showModal: false});
  }

  open() {
    this.setState({showModal: true});
  }

  submit() {
    return fetch('/add_student', {
      method: 'post',
      headers: {'Content-Type':'application/json'},
      body: JSON.stringify({
        name: this.state.name,
        username: this.state.username,
        email: this.state.email,
        notes: this.state.notes
      })
    })
    .then(function(res){
      if (res.status === 200) {
        this.close();
        this.fetchStudentsFromParent();
      } else {
        res.json().then(function(data){
          this.setState({ errors: data.errors });
        }.bind(this));
      }
    }.bind(this));
  }

  handleChange(e) {
    const name = e.target.name;
    const value = e.target.value;
    this.setState({
      [name]: value
    });
  }

  fetchStudentsFromParent() {

  }

  render() {


    return (
        <div className="AddStudentForm">

            <Button className='open-modal'  bsSize="large" active onClick={this.open}>Add Student</Button>
            <Modal className='add-student-modal' show={this.state.showModal} onHide={this.close}>
              <Modal.Header closeButton>
                <Modal.Title>Add Student</Modal.Title>
              </Modal.Header>
              <Modal.Body>
                <ul>
                  { this.state.errors.map(function(error, index){
                    return <li key={index}>{error}</li>;
                  })}
                </ul>
                <form>
                  <FormGroup controlId="name">
                    <ControlLabel>Name: </ControlLabel>
                    <FormControl type="text"
                                 name="name"
                                 value={this.state.name}
                                 onChange={this.handleChange}/>
                  </FormGroup>
                  <FormGroup controlId="username">
                    <ControlLabel>FCC Username: </ControlLabel>
                    <FormControl type="text"
                                 name="username"
                                 value={this.state.username}
                                 onChange={this.handleChange}/>
                  </FormGroup>
                  <FormGroup controlId="email">
                    <ControlLabel>Email: </ControlLabel>
                    <FormControl type="text"
                                 name="email"
                                 value={this.state.email}
                                 onChange={this.handleChange}/>
                  </FormGroup>
                  <FormGroup controlId="Notes">
                    <ControlLabel>Notes: </ControlLabel>
                    <FormControl type="text"
                                 name="notes"
                                 value={this.state.notes}
                                 onChange={this.handleChange}/>
                  </FormGroup>
                  <Button onClick={this.close}>Close</Button>
                  <Button className="submit" onClick={this.submit}>Submit</Button>
                </form>
              </Modal.Body>
            </Modal>
        </div>
    );
  }
}

export default AddStudentForm;