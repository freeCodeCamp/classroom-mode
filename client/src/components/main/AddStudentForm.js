import React, { Component } from 'react'
import {
  Button,
  Modal,
  FormGroup,
  ControlLabel,
  FormControl,
  Alert
} from 'react-bootstrap'
import axios from 'axios'
import './AddStudentForm.css'

const uuidv4 = require('uuid/v4')

const DEFAULT_STATE = {
  showModal: false,
  name: '',
  username: '',
  email: '',
  notes: '',
  errors: [],
}

export default class AddStudentForm extends Component {
  state = DEFAULT_STATE

  close = () => {
    this.setState({ showModal: false })
  }

  open = () => {
    this.setState({ showModal: true })
  }

  submit = () => {
    console.log('submit')
    const { email, notes } = this.state
    return axios
      .post('/students', {
        email,
        notes,
      })
      .then(res => {
        this._fetchStudentsFromParent()
        this.setState({ showModal: !this.state.showModal })
      })
      .catch(e => {
        const { errors } = e.response.data
        this.setState({ errors })
      })
  }

  _fetchStudentsFromParent() {
    this.props.fetchStudentsFromParent()
  }

  handleChange = e => {
    const { name, value } = e.target
    this.setState({
      [name]: value,
    })
  }

  clearErrors = () => this.setState({ errors: [] })

  renderStudentCount() {
    let message
    if (this.props.studentLength > 1) {
      message = `${this.props.studentLength} students`
    } else {
      message = `${this.props.studentLength} student`
    }
    return <div id="student-count">{message}</div>
  }

  render() {
    return (
      <div className="AddStudentForm">
        <div className="class-table-button-container">
          <Button
            className="open-modal"
            bsStyle="success"
            bsSize="large"
            active
            onClick={this.open}
          >
            Add Student
          </Button>
          {this.renderStudentCount()}
        </div>
        <Modal
          className="add-student-modal"
          show={this.state.showModal}
          onHide={this.close}
        >
          <Modal.Header closeButton>
            <Modal.Title>Add Student</Modal.Title>
          </Modal.Header>
          <Modal.Body>
          {this.state.errors &&
            this.state.errors.map(error => (
              <Alert bsStyle="danger" key={error} onDismiss={this.clearErrors}>
                {error}
              </Alert>
            ))
          }
            <form>
              <FormGroup controlId="email">
                <ControlLabel>Email: </ControlLabel>
                <FormControl
                  type="text"
                  name="email"
                  value={this.state.email}
                  onChange={this.handleChange}
                />
              </FormGroup>
              <FormGroup controlId="Notes">
                <ControlLabel>Notes: </ControlLabel>
                <FormControl
                  type="text"
                  name="notes"
                  value={this.state.notes}
                  onChange={this.handleChange}
                />
              </FormGroup>
              <Button onClick={this.close}>Close</Button>
              <Button className="submit" onClick={this.submit}>
                Submit
              </Button>
            </form>
          </Modal.Body>
        </Modal>
      </div>
    )
  }
}
