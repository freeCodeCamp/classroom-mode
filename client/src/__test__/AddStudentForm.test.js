import React from 'react';
import Enzyme, { mount, render, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
Enzyme.configure({ adapter: new Adapter() });
import ReactDOM from 'react-dom';
import AddStudentForm from './../AddStudentForm';
import mockResponse from './mock/response';
import renderer from 'react-test-renderer';

describe("AddStudentForm", () => {
  let props;
  let mountedAddStudentForm;
  const addStudentForm = () => {
    if (!mountedAddStudentForm) {
      mountedAddStudentForm = mount(
        <AddStudentForm {...props} />
      );
    }
    return mountedAddStudentForm;
  }

  beforeEach(() => {
    props = {
      wallpaperPath: undefined,
      userInfoMessage: undefined,
      onUnlocked: undefined,
    };
    mountedAddStudentForm = undefined;
  });

  it("always renders a div", () => {
    const divs = addStudentForm().find("div");
    expect(divs.length).toBeGreaterThan(0);
  });

  it("should open modal when open()", () => {
    addStudentForm().instance().open();
    expect(addStudentForm().instance().state.showModal).toBe(true);
  });

  it("should close modal when close()", () => {
    addStudentForm().instance().setState({showModal: true});
    addStudentForm().instance().close();
    expect(addStudentForm().instance().state.showModal).toBe(false)
  });

  it("should close the modal when submitting returns a 200 response", async () => {
    window.fetch = jest.fn().mockImplementation(function() {
        return Promise.resolve(mockResponse(200, null, '{}'));
      }
    );

    var addStudentFormComponentWrapper = addStudentForm();
    await addStudentFormComponentWrapper.find('button.open-modal').simulate('click');
    await addStudentFormComponentWrapper.find('button.submit').simulate('click');
    
    let modalComponent = addStudentFormComponentWrapper.find('.add-student-modal').get(0);
    
    console.log("*******************************"); 
    console.log("modal at 0: ", modalComponent.props.show); 
    //TODO: check whether modalComponent.props.show has the right value (should be false). 
    //console.log("modal component props: ", modalComponent.props()); 
  });

  it("should returns errors and not close modal when submitting returns a non 200", async () => {
    window.fetch = jest.fn().mockImplementation(function() {
        return Promise.resolve(mockResponse(422, null, JSON.stringify({errors: ["Name is wrong", "Email is wrong"]})));
      }
    );

    var addStudentFormComponentWrapper = addStudentForm();
    await addStudentFormComponentWrapper.find('button.open-modal').simulate('click');
    await addStudentFormComponentWrapper.find('button.submit').simulate('click');
    expect(addStudentFormComponentWrapper.instance().state.showModal).toBe(true);
    expect(addStudentFormComponentWrapper.instance().state.errors[0]).toEqual("Name is wrong");
    expect(addStudentFormComponentWrapper.instance().state.errors[1]).toEqual("Email is wrong");
  });
  
  it("snapshot - should close the modal when submitting returns a 200 response", () => {
    // window.fetch = jest.fn().mockImplementation(function() {
    //     return Promise.resolve(mockResponse(200, null, '{}'));
    //   }
    // );

    let studentForm = renderer.create(<AddStudentForm {...props} />).toJSON();
    // studentForm.find('button.open-modal').simulate('click');
    // studentForm.find('button.submit').simulate('click');
    
    expect(studentForm).toMatchSnapshot();
  })

  it("should set state on handleChange", () => {
    let event = {
      target: {
        name: 'username',
        value: 'newValue'
      }
    }

    let addStudentFormComponentWrapperInstance = addStudentForm().instance();
    addStudentFormComponentWrapperInstance.handleChange(event);
    expect(addStudentFormComponentWrapperInstance.state.username).toEqual('newValue');
  });
});