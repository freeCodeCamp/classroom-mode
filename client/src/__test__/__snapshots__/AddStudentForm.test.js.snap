// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddStudentForm snapshot - should close the modal when submitting returns a 200 response 1`] = `
<div
  className="AddStudentForm"
>
  <button
    className="open-modal btn btn-lg btn-default active"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    Add Student
  </button>
</div>
`;
